### DISCUSSION TITLE
request help: Profile extensions: Is it possible that allow users to add custom code to the Nginx initialization phase?

### DISCUSSION THREAD

**Discussion by griffenliu** [apache/apisix#5029] [Ideas]
### Issue description

我有一些需求，比如在apisix启动时添加一些自定义的初始化逻辑，在apisix管理API的基础上添加一些自定义的管理API等等。想要在Nginx初始化或者Worker初始化阶段做一些事情。是否可以考虑通过配置允许在这两个阶段添加一些自定义代码？
比如：
```
init_by_lua_block {
        require "resty.core"
        apisix = require("apisix")

        local dns_resolver = { {% for _, dns_addr in ipairs(dns_resolver or {}) do %} "{*dns_addr*}", {% end %} }
        local args = {
            dns_resolver = dns_resolver,
        }
        apisix.http_init(args)
        <通过配置添加自定义代码>
    }
    init_worker_by_lua_block {
        apisix.http_init_worker()
        <通过配置添加自定义代码>
    }

```

配置中可以添加：
```
nginx_config: 
  http_init_by_lua_configuration_snippet: |

  http_init_worker_by_lua_configuration_snippet: |
```

### Environment

- apisix version (cmd: `apisix version`):
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

**Comment by tzssangglass** [comment]
> Worker初始化阶段

we can do this in plugin, see: https://github.com/apache/apisix/blob/22ee20c480332ffdb43a1b2e8f4114ea81619d32/apisix/plugins/example-plugin.lua#L62-L68



> 在Nginx初始化

I take a conservative view and I think this is risky. My options are as follows:

- Those who are familiar with the source code will know how to modify it to do this, just as: https://github.com/apache/apisix/blob/22ee20c480332ffdb43a1b2e8f4114ea81619d32/apisix/init.lua#L82
- People who are not familiar with the source code may not be able to use it correctly even if it is open.

So I think this feature somewhat useless.

> nginx_config: 
>   http_init_by_lua_configuration_snippet: |
> 
>   http_init_worker_by_lua_configuration_snippet: |

This is not a good way to do this, we can't write code in yaml. Even if we support i, it would be wrote in the plugin.

@spacewander @tokers @membphis what's your options?

**Reply by spacewander** [reply]
IMHO, it is not a good idea to add Lua snippet to the yaml. It is not a good experience to write Lua in yaml file. What about patching the apisix.http_init like what we have done in the patch.lua?

**Reply by griffenliu** [reply]
What I mean is to set up an extension slot to insert custom code when needed, without having to modify the source code directly.

**Reply by tzssangglass** [reply]
> What I mean is to set up an extension slot to insert custom code when needed, without having to modify the source code directly.

Can you list a scenario of what needs to beadded to the `init_by_lua` phase?

**Reply by griffenliu** [reply]
One of my requirements is that I want to add a custom blacklist plugin. This plugin needs to rely additionally on some custom configuration, cache, etc. support, and also needs to add some interfaces for management functions similar to apisix. Our approach is to mimic apisix by defining a set of our own, which requires introducing `apisix`-like global variables in the init phase and initializing the relevant resources in the init_worker phase.

**Reply by spacewander** [reply]
Nginx conf is conf so it can be put into the yaml conf.
Put Lua code to conf doesn't work well when your code is growing.

**Reply by griffenliu** [reply]
My idea was that I could insert code like this:
```
init_by_lua_block {
        require "resty.core"
        apisix = require("apisix")

        local dns_resolver = { {% for _, dns_addr in ipairs(dns_resolver or {}) do %} "{*dns_addr*}", {% end %} }
        local args = {
            dns_resolver = dns_resolver,
        }
        apisix.http_init(args)

        my_module = require('my_module')
        my_module.http_init()
    }

    init_worker_by_lua_block {
        apisix.http_init_worker()
        my_module.http_init_worker()
    }
```
This way I can use my own lua file to handle some additional operations.

**Reply by griffenliu** [reply]
> Nginx conf is conf so it can be put into the yaml conf.
> Put Lua code to conf doesn't work well when your code is growing.



> IMHO, it is not a good idea to add Lua snippet to the yaml. It is not a good experience to write Lua in yaml file. What about patching the apisix.http_init like what we have done in the patch.lua?

Or is there a sample example to refer to when you say you use patch.lua? Where can I find patch.lua and how does it work?

**Reply by tzssangglass** [reply]
patch.lua: https://github.com/apache/apisix/blob/master/apisix/patch.lua

**Reply by griffenliu** [reply]
Does this still require modifying the source code of apisix itself? Is there an entry point for external developers to add patches like this themselves? My intention is not to modify the source code of apisix, we were using version 2.2 and modified a lot of source code. apisix is growing so fast and iterating so fast that version upgrade is a problem.

**Reply by tzssangglass** [reply]
I got. Perhaps we should discuss how to open the `init_by_lua` phase. cc @spacewander @membphis

**Reply by spacewander** [reply]
I think we should provide a way to specify a custom "patch.lua" or a "hook.lua". In this user-provided file, you can inject the code into the phase of APISIX like this: https://github.com/apache/apisix/blob/cd97d827abf6791611ce355b12555531dbd7b53d/apisix/patch.lua#L312.

BTW, open the `init_by_lua` doesn't mean your code is always compatible with APISIX. We don't promise any compatibility with the Nginx configuration:

https://github.com/apache/apisix/blob/cd97d827abf6791611ce355b12555531dbd7b53d/conf/config-default.yaml#L179-L182

It is impossible to give this promise with Nginx configuration, and it is more impossible to give it with Lua code.

**Reply by griffenliu** [reply]
I can understand that, but it's much easier to modify your own code than modifying third-party code.

**Reply by griffenliu** [reply]
I think the first option is also feasible, I just want a way to inject my own code in these phases, either by adding my own patch.lua or by adding custom code in the init phase.
